src/Moq/Mock.cs

```diff
@@ -259,42 +259,77 @@ private MockException TryVerifySetups(Func<Setup, MockException> verifySetup)
 			}
 		}
 
-		internal static void VerifyVoid(Mock mock, LambdaExpression expression, Times times, string failMessage)
+		internal static void Verify(Mock mock, LambdaExpression expression, Times times, string failMessage)
 		{
 			Guard.NotNull(times, nameof(times));
 
-			var (obj, method, args) = expression.GetCallInfo(mock);
-			ThrowIfVerifyExpressionInvolvesUnsupportedMember(expression, method);
-
-			var expectation = new InvocationShape(method, args);
-			VerifyCalls(GetTargetMock(obj, mock), expectation, expression, times, failMessage);
+			var parts = expression.Split();
+			Mock.VerifyRecursive(mock, expression, parts, times, failMessage, verifyLast: (part, targetMock) =>
+			{
+				var expectation = new InvocationShape(part.Method, part.Arguments);
+				VerifyCalls(targetMock, expectation, expression, times, failMessage);
+			});
 		}
 
-		internal static void VerifyNonVoid(Mock mock, LambdaExpression expression, Times times, string failMessage)
+		internal static void VerifyGet(Mock mock, LambdaExpression expression, Times times, string failMessage)
 		{
 			Guard.NotNull(times, nameof(times));
 
-			if (expression.IsProperty())
+			if (!expression.IsPropertyIndexer())  // guard because `.ToPropertyInfo()` doesn't (yet) work for indexers
 			{
-				VerifyGet(mock, expression, times, failMessage);
+				var property = expression.ToPropertyInfo();
+				Guard.CanRead(property);
 			}
-			else
-			{
-				var (obj, method, args) = expression.GetCallInfo(mock);
-				ThrowIfVerifyExpressionInvolvesUnsupportedMember(expression, method);
 
-				var expectation = new InvocationShape(method, args);
-				VerifyCalls(GetTargetMock(obj, mock), expectation, expression, times, failMessage);
-			}
+			var parts = expression.Split();
+			Mock.VerifyRecursive(mock, expression, parts, times, failMessage, verifyLast: (part, targetMock) =>
+			{
+				var expectation = new InvocationShape(part.Method, part.Arguments);
+				VerifyCalls(targetMock, expectation, expression, times, failMessage);
+			});
```

Comment:
There's some code duplication here. We might just as well delegate to `Mock.Verify`.